### PR TITLE
Remove Image settings page and standardise Paperclip styles

### LINKED ISSUE #5695: Remove unchangable config settings from Configuration
## Description
Image settings are not something that can be changed without the help of a dev: running commands on the server, moving images around, etc.
I think we can probably delete completely this page:
image

The only part I have doubts are the paperclips styles, can this be changed? Will they have an impact anywhere on the app? Are these configs useful?

## Expected Behavior
No image settings in Configuration.

## Actual Behaviour
Image settings can be changed and images in the app will break.

## Steps to Reproduce
There are few cases:
- change attachments URL and see that all existing images are now gone.
- change use s3 for images and see that all images are now gone.

## Severity
I think this is the cause of S1 #5668
bug-s1: a critical feature is broken: checkout, payments, signup, login

## Your Environment
Seen in BE prod.

### PR DESCRIPTION
**Pull request by luisramos0**
#### What? Why?

Closes #5695

This PR is paired with changes in ofn-install to set and change these setting from now on: https://github.com/openfoodfoundation/ofn-install/pull/645

Apart from removing the pages we are adding some code to spree initializers to pick up ofn-instrall config if they exist.

#### What should we test?
After this PR and after image styles were removed from ofn-install as part of https://github.com/openfoodfoundation/ofn-install/pull/645 there is no way to update styles though the UI or through ofn-install.
We end up with only two ways to update styles:
- changing the default in the code like in this PR: https://github.com/openfoodfoundation/openfoodnetwork/pull/6223 
- use the new tools: https://github.com/openfoodfoundation/openfoodnetwork/pull/6014

This PR becomes a manual test only:
- image settings are gone from configuration
- images are rendering correctly as before this PR

#### Release notes
Changelog Category: Removed
Removed image settings page from the app, image settings are now done in ofn-isntall.

### CODE REVIEW

File: knapsack_rspec_report.json
Comment (sauloperez) [inline knapsack_rspec_report.json:334]: I think this will unbalance our jobs. We might need a full re-balance and update this file.

File: config/initializers/spree.rb
Comment (sauloperez) [inline config/initializers/spree.rb:90]: What about bringing in the image model from spree and define the attachments as usual for Paperclip? The first place I'd go find the settings is in the model because that's the convention. Also, I find tedious having to deal with hashes here instead of the `has_attached_file` DSL, and is not easy to follow either IMO.

File: knapsack_rspec_report.json
Comment (luisramos0) [reply knapsack_rspec_report.json:334]: yes, but on another PR. I havent managed to get a new knapsack report the last time I tried, the single job build expired after 3 hours :man_shrugging:

File: config/initializers/spree.rb
Comment (luisramos0) [reply config/initializers/spree.rb:90]: I'd have to investigate how do it, change the code and test it.
This way I am keeping the code as is (this is exactly what was on the controller) and just moving it around. I see changing code a bit out of scope for this PR. It would be more effort and risk. I'd prefer to go forward as is so that we save dev time. What do you think?

File: knapsack_rspec_report.json
Comment (sauloperez) [reply knapsack_rspec_report.json:334]: sure, I didn't mean to do it here.

**Comment by sigmundpetersen** [comment]
Some conflicts that should be resolved @luisramos0

**Review by mkllnk — APPROVED** [review]
I reviewed this pull request a while ago but somehow it wasn't submitted. My draft was still sitting here. I commented on one thing that I think we should change but now I think that we should get this through and follow up with another pull request.

Since we set the same styles for all instances, I think we should hardcode them in the Spree::Image model. There is no use case for having different styles in different countries. We all use the same design and if someone finds an optimisation, we should share it with everybody.

There's another issue I came across: When we deploy this, the new styles will come into effect. But until the whole thumbnail generation is complete, there will be broken images. That's not due to different sizes but because of different file names (jpg instead of jpeg and png). The only way around that is to run the thumbnail generation with new styles before deploying this.

The rollout of this is driving me a bit crazy. It's not that hard when you look at it calmly but I want to have it all done at the same time and there are quite a few steps. ;-)

**Comment by mkllnk** [comment]
### Testing notes


-    verify we can change image settings through ofn-install (testing PR 64) :heavy_check_mark: 
-    verify those changes are correctly picked up in the application :red_circle: 

The effective style was not changed. All styles should have `jpg` as format:

```diff
- "{\"mini\":[\"48x48#\",\"jpg\"],\"small\":[\"227x227#\",\"jpg\"],\"product\":[\"240x240\\u003E\",\"\"],\"large\":[\"600x600\\u003E\",\"jpg\"]}"
+ "{\"mini\":[\"48x48#\",\"jpg\"],\"small\":[\"227x227#\",\"jpg\"],\"product\":[\"240x240>\",\"jpg\"],\"large\":[\"600x600>\",\"jpg\"]}"
```

But there is no change.

Afterwards we should also do a manual verification that:

-    image settings are gone from configuration :heavy_check_mark: 
-    images are rendering correctly as before this PR :red_circle: 

The attachment_path was changed which breaks all images:
```diff
- public/images/spree/products/:id/:style/:basename.:extension
+ public/spree/products/:id/:style/:basename.:extension
```

**Comment by sigmundpetersen** [comment]
Hey @luisramos0 are the build failures unrelated to the PR?
They seem consistent.

**Review by mkllnk — APPROVED** [review]
Yay, we are getting there! :tada: 

Really good work, finally ditching Spree::Config which has annoyingly cached values.

My last dev test had two fails. I think that the first one was caused by the cache which should be resolved by the new commits here. The second fail was caused by a missing attachment path config for au-staging. I opened https://github.com/openfoodfoundation/ofn-install/pull/658 to fix that. The dev test of this PR is blocked by that config.

File: app/models/spree/image.rb
Comment (luisramos0) [reply app/models/spree/image.rb:15]: Don't we need to deploy this immediately after we convert the styles with the task?

File: app/models/spree/image.rb
Comment (luisramos0) [reply app/models/spree/image.rb:60]: :+1:

File: app/models/spree/image.rb
Comment (mkllnk) [reply app/models/spree/image.rb:15]: Yes, so #6014 needs to be deployed already. It could avoid delays if we run the scripts in #6014 early then run it again with the new images just before we want to deploy this here.

File: app/models/spree/image.rb
Comment (sauloperez) [inline app/models/spree/image.rb:66]: We should replace all these `Spree::Image` with `self`

**Review by sauloperez — CHANGES_REQUESTED** [review]
yay, so much :fire: !

File: app/models/spree/image.rb
Comment (sauloperez) [inline app/models/spree/image.rb:56]: we can extract that into a method to make this a bit more readable. Code Climate complaints about it too

File: app/models/spree/image.rb
Comment (luisramos0) [reply app/models/spree/image.rb:66]: nice :+1: actually it can be removed because then self becomes redundant (rubocop helps remembering that).

File: app/models/spree/image.rb
Comment (luisramos0) [reply app/models/spree/image.rb]: :+1:

**Review by sauloperez — APPROVED** [review]

**Comment by luisramos0** [comment]
The restart checkout spec is failing in semaphore but passing locally. I dont see how this PR could be related to it...

**Comment by sigmundpetersen** [comment]
Interestingly the same specs fail in another PR as well https://semaphoreci.com/openfoodfoundation/openfoodnetwork-2/branches/pull-request-6110/builds/1 ping @filipefurtad0 

So I guess that also indicates it's not specific to this PR. I haven't seen it fail in master though. Let's see if it will start failing after merging this 🙈

**Comment by luisramos0** [comment]
good point @sigmundpetersen thanks
I think we can go ahead then and have a look at it if it starts failing in master.

**Comment by filipefurtad0** [comment]
Thanks for pinging @sigmundpetersen :+1: 
I rebuilt the other PR6110 and it failed again. Also, it looks like `/spec/services/order_checkout_restart_spec.rb` it's failing in master [now](https://semaphoreci.com/openfoodfoundation/openfoodnetwork-2/branches/master/builds/1751)...

**Comment by Matt-Yorkley** [comment]
@luisramos0 there's two big changes here and I'm wondering if we can decouple them?

- removing the admin page for image settings
- changing the image sizes as part of the product images changes (which requires the image resizing to be done before this could be deployed?)

**Comment by luisramos0** [comment]
Hey @Matt-Yorkley the change in image.rb has_attached_file :attachment will not have any effect unless we run the commands to resize the images. Are you suggesting I create a separate PR just for that line in image.rb has_attached_file :attachment?

**Comment by luisramos0** [comment]
ok @Matt-Yorkley I have rebased this PR and added a commit to remove the change in the default styles.

**Review by Matt-Yorkley — APPROVED** [review]

**Comment by sigmundpetersen** [comment]
Hey @luisramos0 is this still blocked by #6014 or is it rather #6223 that is blocked?
(Or nothing blocked anymore? 🙂 )

**Comment by sauloperez** [comment]
As far as I understand nothing is blocked anymore.

**Comment by luisramos0** [comment]
yes, this is not blocked by anything now :+1: that was the point of the change.
But #6223 is very closely related to whatever we do in #6014

**Comment by luisramos0** [comment]
So this one is ready for manual testing now :+1:

**Comment by mkllnk** [comment]
I think that we are going in circles here. The breaking change in this PR is that the image settings are not read from the database any more and instead defined in the code. The current settings in the production databases are not standardised and any standardised setting would break some servers. The Australian server converts all product images to JPG while other servers keep them in their original format which is often PNG and sometimes JPEG. After deploying the current version of this pull request (using original image formats), Australia would suddenly lose all PNG and JPEG images. Because of the style declarations, the code would look for a PNG image while we only stored the JPG image.

So this still depends on the tools to convert the images first and the last commit doesn't bring any improvement.

**Comment by Matt-Yorkley** [comment]
> Australia would suddenly lose all PNG and JPEG images.

I don't think it will. The image format forces uploaded images to be created in that format. If you don't specify a format it will still find whatever images have been saved previously.

**Comment by Matt-Yorkley** [comment]
If we use the jpg format eg: `["227x227#","jpg"]` it will break images on all servers except Aus (and Germany).

If we leave out the jpg format eg: `"227x227#"` it will not break any images on any servers.

**Comment by mkllnk** [comment]
I still disagree with that one:
> If we leave out the jpg format eg: small: "227x227#" it will not break any images on any servers.

It's difficult to focus the conversation on all the different pull requests. Maybe let's continue on Slack?

**Comment by Matt-Yorkley** [comment]
I've done a `dev-test` here. I changed the image settings in the UI to something other than the ones in this PR before staging.

- Is the image setting page removed? :heavy_check_mark: 
- Are existing product images still shown? :heavy_check_mark: 
- Are the hard-coded image settings in `Spree::Image` now the single source of truth for image configs, and the old database-persisted configs are no longer used? :heavy_check_mark: 
- Does image upload work for products, and apply the expected formatting? :heavy_check_mark:

**Comment by luisramos0** [comment]
thanks Matt!

**Pull request merged**
